include/cetl/pmr/function.hpp

```diff
@@ -0,0 +1,359 @@
+
+    /// @brief Moves the target of `other` to `*this`
+    ///
+    /// Any failure during the movement will result in the "valueless by exception" state.
+    ///
+    function& operator=(function&& other) noexcept
+    {
+        if (this != &other)
+        {
+            any_handler_       = std::move(other.any_handler_);
+            handler_ptr_       = get_if<handler_t>(&any_handler_);
+            other.handler_ptr_ = nullptr;
+        }
+        return *this;
+    }
+
+    /// @brief Drops the current target. `*this` is empty after the call.
+    ///
+    function& operator=(std::nullptr_t) noexcept
+    {
+        reset();
+        return *this;
+    }
+
+    /// @brief Checks whether `*this` stores a callable function target, i.e. is not empty.
+    ///
+    explicit operator bool() const noexcept
+    {
+        CETL_DEBUG_ASSERT(any_handler_.has_value() == (nullptr != handler_ptr_), "");
+
+        return any_handler_.has_value();
+    }
+
+    /// @brief Invokes the stored callable function target with the parameters args.
+    ///
+    /// @param args Arguments to pass to the stored callable function target.
+    /// @return None if `Result` is `void`. Otherwise the return value of the invocation of the stored callable object.
+    /// @throws `std::bad_function_call` if `*this` does not store a callable function target, i.e. `!*this == true`.
+    ///
+    Result operator()(Args... args) const
+    {
+        if (!any_handler_.has_value())
+        {
+            detail::throw_bad_function_call();
+        }
+
+        CETL_DEBUG_ASSERT(nullptr != handler_ptr_, "");
+
+        return handler_ptr_->operator()(args...);
+    }
+
+    /// @brief Resets the current target. `*this` is empty after the call.
+    ///
+    void reset() noexcept
+    {
+        any_handler_.reset();
+        handler_ptr_ = nullptr;
+    }
+
+    /// @brief Resets the current target, and sets specific memory resource. `*this` is empty after the call.
+    ///
+    template <typename PmrAlias = Pmr, typename = cetl::detail::EnableIfPmrT<PmrAlias>>
+    void reset(Pmr* const mem_res) noexcept
+    {
+        any_handler_.reset(mem_res);
+        handler_ptr_ = nullptr;
+    }
+
+    /// @brief Swaps the contents of `*this` and `other`.
+    ///
+    /// Any failure during the swap operation could result in the "valueless by exception" state,
+    /// and depending on which stage of swapping the failure happened
+    /// it could affect (invalidate) either of `*this` or `other` function.
+    /// Use `valueless_by_exception()` method to check if a function is in such failure state,
+    /// and `reset` (or `reset(Pmr*)`) method to recover from it.
+    ///
+    void swap(function& other) noexcept
+    {
+        any_handler_.swap(other.any_handler_);
+
+        handler_ptr_       = get_if<handler_t>(&any_handler_);
+        other.handler_ptr_ = get_if<handler_t>(&other.any_handler_);
+    }
+
+    /// True if the function is valueless b/c of an exception.
```

Comment:
```suggestion
    /// True if the function is valueless b/c of an exception or OOM.
```

OOM doesn't necessarily cause an exception